server/src/main/java/org/elasticsearch/ingest/IngestSourceAndMetadata.java

```diff
@@ -0,0 +1,613 @@
+        IngestDocument.Metadata.DYNAMIC_TEMPLATES.getFieldName(),
+        IngestSourceAndMetadata::mapValidator,
+        IngestDocument.Metadata.IF_SEQ_NO.getFieldName(),
+        IngestSourceAndMetadata::longValidator,
+        IngestDocument.Metadata.IF_PRIMARY_TERM.getFieldName(),
+        IngestSourceAndMetadata::longValidator,
+        IngestDocument.Metadata.TYPE.getFieldName(),
+        IngestSourceAndMetadata::stringValidator
+    );
+
+    protected final Map<String, Object> source;
+    protected final Map<String, Object> metadata;
+    protected final Map<String, BiFunction<String, Object, Object>> validators;
+    private EntrySet entrySet; // cache to avoid recreation
+
+    IngestSourceAndMetadata(
+        String index,
+        String id,
+        long version,
+        String routing,
+        VersionType versionType,
+        ZonedDateTime timestamp,
+        Map<String, Object> source
+    ) {
+        this(new HashMap<>(source), metadataMap(index, id, version, routing, versionType), timestamp, VALIDATORS);
+    }
+
+    /**
+     * Creates an {@code IngestSourceAndMetadata} from the given source, metadata and timestamp
+     * @param source the source document map
+     * @param metadata the metadata map
+     * @param timestamp the time of ingestion
+     */
+    IngestSourceAndMetadata(Map<String, Object> source, Map<String, Object> metadata, ZonedDateTime timestamp) {
+        this(source, metadata, timestamp, VALIDATORS);
+    }
+
+    /**
+     * Create IngestSourceAndMetadata with custom validators.
+     *
+     * @param source the source document map
+     * @param metadata the metadata map
+     * @param timestamp the time of ingestion
+     * @param validators validators to run on metadata map, if a key is in this map, the value is stored in metadata
+     */
+    IngestSourceAndMetadata(
+        Map<String, Object> source,
+        Map<String, Object> metadata,
+        ZonedDateTime timestamp,
+        Map<String, BiFunction<String, Object, Object>> validators
+    ) {
+        this.source = source != null ? source : new HashMap<>();
+        this.metadata = metadata != null ? metadata : new HashMap<>();
+        this.timestamp = timestamp;
+        this.validators = validators;
+        validateMetadata();
+    }
+
+    /**
+     * Create a IngestSourceAndMetadata using the underlying map and set of validators.  The validators are applied to the map to ensure
+     * the incoming map matches the invariants enforced by the validators.
+     * @param sourceAndMetadata the wrapped map.  Should not be externally modified after creation.  This map is modified to remove
+     *                          metadata values and will become source
+     * @param timestamp the timestamp of ingestion
+     * @throws IllegalArgumentException if a validator fails for a given key
+     */
+    public static IngestSourceAndMetadata ofMixedSourceAndMetadata(Map<String, Object> sourceAndMetadata, ZonedDateTime timestamp) {
```

Comment:
Similar to IngestDocument, there seem to be a lot of different ways to construct an IngestSourceAndMetadata. Can these be consolidated?

Reply:
Reduced to two.